test/apps/iodemo/io_demo.cc

```diff
@@ -202,19 +236,65 @@ class IoDemoRandom {
         return size;
     }
 
+    static void setMemoryType(ucs_memory_type_t memory_type)
+    {
+        _memory_type = memory_type;
+    }
+
 private:
     template <typename T>
     static inline void fill(unsigned &seed, T *buffer, size_t count) {
+#ifdef HAVE_CUDA
+        T temp;
+#endif
+
         for (size_t i = 0; i < count; ++i) {
-            buffer[i] = rand<T>(seed);
+            switch (_memory_type) {
+#ifdef HAVE_CUDA
+            case UCS_MEMORY_TYPE_CUDA:
+                temp = rand<T>(seed);
+                cudaMemcpy(&buffer[i], &temp, sizeof(T), cudaMemcpyDefault);
```

Comment:
maybe add cuda kernel to calculate rand on GPU to avoid Memcpy? (the same could be for validation)

Reply:
Will be in a separate PR.